### PR TITLE
bpf: correctly encapsulate pod to node traffic with kube-proxy+hostfw

### PR DESCRIPTION
**Pull request by giorio94**
When the host firewall is enabled in tunneling mode, pod to node traffic needs to be forwarded through the tunnel in order to preserve the security identity (as otherwise the source IP address would be SNATted), which is required to enforce ingress host policies.

One tricky case is represented by node (or hostns pod) to pod traffic via services with local ExternalTrafficPolicy, when KPR is disabled. Indeed, in this case, the SYN packet is routed natively (as both the source and the destination are node IPs) to the destination node, and then DNATted to one of the backend IPs, without being SNATted at the same time. Yet, the SYN+ACK packet would then be incorrectly redirected through the tunnel (as the destination is a node IP, associated with a tunnel endpoint in the ipcache), hence breaking the connection, while it should be passed to the stack to be rev DNATted and then forwarded accordingly.

In detail, reporting the description from c8052a1fab8b, the broken packet path is node1 --VIP--> pod@node2 (VIP is node2IP):

- SYN leaves node1 via native device with  node1IP -> VIP
- SYN is DNATed on node2 to                node1IP -> podIP
- SYN is delivered to lxc device with      node1IP -> podIP
- SYN+ACK is sent from lxc device with     podIP   -> node1IP
- SYN+ACK is redirected in BPF directly to cilium_vxlan
- SYN+ACK arrives on node1 via tunnel with podIP   -> node1IP
- RST is sent because podIP doesn't match VIP

c8052a1fab8b attempted to fix this issue for the kube-proxy+hostfw (and IPSec) scenarios by always passing the packets to the stack, so that it doesn't bypass conntrack. The IPSec specific workaround got then removed in 0a8f2c4ee43e, as that path asymmetry is no longer present. However, always passing packets to the stack breaks the host firewall policy enforcement for pod to node traffic, as at that point there's no route which redirects these packets back to the tunnel to preserve the security identity, and they get simply masqueraded and routed natively.

To prevent this issue, let's pass packets to the stack only if they are a reply with destination identity matching a remote node, as in that case they may need to be rev DNATted. There are two possibilities at that point: (a) the destination is a CiliumInternalIP address, and the reply needs to go through the tunnel -- node routes ensure that the packet is first forwarded to cilium_host, before being redirected through the tunnel; (b) the destination is one of the other node
addresses, and the reply needs to be forwarded natively according to the local routing table (as node to pod/node traffic never goes through the tunnel unless the source is a CiliumInternalIP address).

Overall, this change addresses the externalTrafficPolicy=local service case, while still preserving encapsulation in all other cases. As a side effect, it also improves the performance in the kube-proxy + hostfw case, as pod to pod traffic gets now also redirected immediately through the tunnel, instead of being sent via the stack.

Fixes: c8052a1fab8b ("bpf: Do not bypass conntrack if running kube-proxy+hostfw or IPSec")

```release-note
Fix host firewall policy enforcement for pod to node traffic when tunneling is enabled and KPR is disabled
```

### CODE REVIEW

**Comment by giorio94** [comment]
Force pushed to clarify the commit message around the possibility of traffic with destination identity matching a remote node needing to go through the tunnel if the destination is a CiliumInternalIP. Marking ready for review again.

File: bpf/bpf_lxc.c [hunk 1]
```diff
@@ -1192,6 +1198,20 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
...
+		/*
+		 * For the host firewall, traffic from a pod to a remote node is sent
+		 * through the tunnel. In the case of node to remote pod traffic via
+		 * externalTrafficPolicy=Local services, packets may be DNATed when
+		 * they enter the remote node (without being SNATed at the same time).
+		 * If kube-proxy is used, the response needs to go through the stack
+		 * to apply the correct reverse DNAT, and then be routed accordingly.
+		 * See #14674 for details.
+		 */
+		if (ct_status == CT_REPLY && identity_is_remote_node(*dst_sec_identity))
```
Comment (julianwiedmann) [inline bpf/bpf_lxc.c]: drive-by - strictly speaking, I believe we might also need to handle `CT_RELATED` here:
https://github.com/cilium/cilium/blob/a9d83fe157e14bfa10a1f3437bc0c4f98a5c800f/bpf/lib/conntrack.h#L823-L827

And yes, wouldn't it be nice if we had a `ct_result_is_reply()` helper for such cases :slightly_smiling_face: .

Comment (giorio94) [reply bpf/bpf_lxc.c]: TIL! Thanks, fixed.

File: bpf/bpf_lxc.c [hunk 2]
```diff
@@ -650,6 +650,13 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 		key.ip6.p4 = 0;
```
Comment (pchaigno) [inline bpf/bpf_lxc.c:650]: > However, always passing packets to the stack breaks the host firewall policy enforcement for pod to node traffic, as at that point there's no route which redirects these packets back to the tunnel to preserve the security identity, and they get simply masqueraded and routed natively.
> 
> To prevent this issue, let's pass packets to the stack only if they are a reply with destination identity matching a remote node, as in that case they may need to be rev DNATted.

I'm a bit confused here. So passing packets to the stack means they are not sent through the tunnel and we lose the source identity. You propose to be more selective and only send reply packets to remote nodes to the stack. I understand that will limit the impact (less cases where we lose the source identity), but it won't completely fix the issue. Replies from a pod to a remote node will still lose their source identity. What am I missing?

**Review by pchaigno — APPROVED** [review]
Thanks for tackling this! It's been known broken since forever :disappointed: 

EDIT: I meant to comment instead of approving :shrug:

File: bpf/bpf_lxc.c (hunk 2 above)
Comment (giorio94) [reply bpf/bpf_lxc.c:650]: After thinking a bit more about this (and doing some tests), my understanding is that node to pod traffic would not be affected. AFAIU, there are two possible cases depending on the source address:

* CiliumInternalIP -> remote pod: the packet is redirected to `cilium_host` by the route covering the remote PodCIDR, processed by `cil_from_host`, forwarded through the tunnel and eventually to the remote pod. The reply would then be processed by `cil_from_container`, and passed to the stack as per the above changes (given that the destination identity is `remote_node`). At that point, it would be redirected to `cilium_host` (as the CiliumInternalIP belongs to the remote PodCIDR), forwarded through the tunnel (with the correct source identity, [configured by `cil_from_host`](https://github.com/giorio94/cilium/blob/mio%2Fhost-firewall-tunnel/bpf/bpf_host.c#L1134-L1135)), eventually reaching the origin node again. The source IP is also preserved, so the identity could be looked up on the destination node.
* InternalIP -> remote pod: the packet is redirected to `cilium_host` as before, source natted to the CiliumInternalIP address (third rule below) and then continue along the lines of the previous case. 
   ```
  Chain CILIUM_POST_nat (1 references)
  target     prot opt source               destination
  MASQUERADE  all  --  10.244.2.0/24       !10.244.2.0/24        /* cilium masquerade non-cluster */
  ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            mark match 0xa00/0xe00 /* exclude proxy return traffic from masquerade */
  SNAT       all  -- !10.244.2.0/24       !10.244.2.0/24        /* cilium host->cluster masquerade */ to:10.244.2.118
  SNAT       all  --  127.0.0.1            0.0.0.0/0            /* cilium host->cluster from 127.0.0.1 masquerade */ to:10.244.2.118
  SNAT       all  --  0.0.0.0/0            0.0.0.0/0            mark match 0xf00/0xf00 ctstate DNAT /* hairpin traffic that originated from a local pod */ to:10.244.2.118
   ```

I guess we could be even more specific in limiting the packets passed to the stack, and do that only if the original packet didn't come from the tunnel (`!ct_state->from_tunnel`). My understanding, though, is that at the moment it gets set [only in the IPv4 path](https://github.com/giorio94/cilium/blob/mio%2Fhost-firewall-tunnel/bpf/bpf_lxc.c#L1971), and would require some additional changes to make it work for IPv6 as well, making this fix a little harder to backport (depending on if and to which stable branches we would backport it).

Comment (pchaigno) [reply bpf/bpf_lxc.c:650]: I'm not worried about the case `node -> remote pod`. That doesn't need tunneling to preserve the source identity since the source identity is `REMOTE_NODE_ID` anyway.

I'm worried about this case:
Assuming a connection `node1 --VIP--> pod@node2` (VIP is node2IP).
- SYN leaves node1 via native device with                     node1IP -> VIP
- SYN is DNATed on node2 to                                        node1IP -> podIP
- SYN is delivered to lxc device with                               node1IP -> podIP
- SYN+ACK is sent from lxc device with                         podIP -> node1IP
- SYN+ACK is sent to the stack because it's a reply packet and the destination is a node IP.
- SYN+ACK is reverse DNATed to                                   VIP -> node1IP
- SYN+ACK arrives on node1 via the native device with VIP -> node1IP

When the SYN+ACK packet reaches node1, we lost its source identity (i.e., pod@node2).

That said, now that I write this down, maybe it's actually fine because we don't need the source identity on reply traffic given we already rely on CT for policy enforcement? Could we need it for anything other than policy enforcement? Hubble I guess...

Comment (giorio94) [reply bpf/bpf_lxc.c:650]: > Assuming a connection node1 --VIP--> pod@node2 (VIP is node2IP).

I assume you are talking about the ExternalTrafficPolicy=Local case here.

> When the SYN+ACK packet reaches node1, we lost its source identity (i.e., pod@node2).

I wonder which identity would be even correct in this case, as from the PoW of node1, the destination identity would be remote_node (as the dst IP is that of the node), and not the pod identity. And it would be matched by a CCNP which allows egress traffic towards the node, not the pod. So it sounds reasonable to me that also the reply would have a source identity corresponding to the node.

My understanding is that the same would also apply in case of a service with ExternalTrafficPolicy=Cluster. A special case would then be `node1 --> VIP@node2 --> pod@node3`. The `node1->node2` path would be outside of the tunnel (with dst -- src for the reply -- identity remote_node), while `node2->node3` inside (given the SNATting performed on node2), (with dst -- src for the reply -- identity pod).

Comment (pchaigno) [reply bpf/bpf_lxc.c:650]: > I assume you are talking about the ExternalTrafficPolicy=Local case here.

Yep.

> I wonder which identity would be even correct in this case, as from the PoW of node1, the destination identity would be remote_node (as the dst IP is that of the node), and not the pod identity. And it would be matched by a CCNP which allows egress traffic towards the node, not the pod. So it sounds reasonable to me that also the reply would have a source identity corresponding to the node.

Hm, maybe you're right. I didn't see things from that angle.

**Review by pchaigno — APPROVED** [review]
LGTM!

Is there an end-to-end test coming to cover this?

**Comment by giorio94** [comment]
> Is there an end-to-end test coming to cover this?

Yep, I've opened https://github.com/cilium/cilium-cli/pull/2306 to introduce basic host firewall connectivity tests as part of the Cilium CLI, and once merged I'll update the conformance-e2e workflow to enable host firewall also for the KPR=disabled case (either as part of this PR or a separate one, as you prefer).

Do you have opinions on backporting this change to stable branches? I've currently marked it for v1.15, but I wonder whether it could make sense to backport it to older ones as well.

**Comment by pchaigno** [comment]
> Do you have opinions on backporting this change to stable branches? I've currently marked it for v1.15, but I wonder whether it could make sense to backport it to older ones as well.

I wouldn't backport further back for now, unless users ask for it. We can always backport later if needed and, in the meantime, we gain some confidence that everything works as expected.

**Comment by giorio94** [comment]
I've opened https://github.com/cilium/cilium/pull/30914 to switch two existing Conformance E2E configurations (with KPR disabled) to also enable host firewall. I couldn't add a commit to this PR as the branch is from a fork.

**Comment by giorio94** [comment]
CI is green, reviews are in, marking ready to merge.
